proxy/src/serverless.rs

```diff
@@ -90,51 +92,84 @@ pub async fn task_main(
     tls_server_config.alpn_protocols = vec![b"h2".to_vec(), b"http/1.1".to_vec()];
     let tls_acceptor: tokio_rustls::TlsAcceptor = Arc::new(tls_server_config).into();
 
-    let mut addr_incoming = AddrIncoming::from_listener(ws_listener)?;
-    let _ = addr_incoming.set_nodelay(true);
-    let addr_incoming = ProxyProtocolAccept {
-        incoming: addr_incoming,
-        protocol: "http",
-    };
-
     let ws_connections = tokio_util::task::task_tracker::TaskTracker::new();
     ws_connections.close(); // allows `ws_connections.wait to complete`
 
-    let tls_listener = TlsListener::new(tls_acceptor, addr_incoming, config.handshake_timeout);
-
-    let make_svc = hyper::service::make_service_fn(
-        |stream: &tokio_rustls::server::TlsStream<
-            WithConnectionGuard<WithClientIp<AddrStream>>,
-        >| {
-            let (conn, _) = stream.get_ref();
-
-            // this is jank. should dissapear with hyper 1.0 migration.
-            let gauge = conn
-                .gauge
-                .lock()
-                .expect("lock should not be poisoned")
-                .take()
-                .expect("gauge should be set on connection start");
-
-            // Cancel all current inflight HTTP requests if the HTTP connection is closed.
-            let http_cancellation_token = CancellationToken::new();
-            let cancel_connection = http_cancellation_token.clone().drop_guard();
-
-            let span = conn.span.clone();
-            let client_addr = conn.inner.client_addr();
-            let remote_addr = conn.inner.inner.remote_addr();
-            let backend = backend.clone();
-            let ws_connections = ws_connections.clone();
-            let endpoint_rate_limiter = endpoint_rate_limiter.clone();
-            let cancellation_handler = cancellation_handler.clone();
+    let http_connections = tokio_util::task::task_tracker::TaskTracker::new();
+    http_connections.close();
+
+    let server = hyper_util::server::conn::auto::Builder::new(hyper_util::rt::TokioExecutor::new());
+
+    while let Some(res) = run_until_cancelled(ws_listener.accept(), &cancellation_token).await {
+        let (conn, peer_addr) = res.context("could not accept TCP stream")?;
+        if let Err(e) = conn.set_nodelay(true) {
+            tracing::error!("could not set nodolay: {e}");
+            continue;
+        }
+        let conn_id = uuid::Uuid::new_v4();
+        let http_conn_span = tracing::info_span!("http_conn", ?conn_id);
+        let session_id = uuid::Uuid::new_v4();
+        let cancellation_token = cancellation_token.child_token();
+
+        let tls = tls_acceptor.clone();
+
+        let backend = backend.clone();
+        let ws_connections = ws_connections.clone();
+        let endpoint_rate_limiter = endpoint_rate_limiter.clone();
+        let cancellation_handler = cancellation_handler.clone();
+        let server = server.clone();
+
+        http_connections.spawn(
```

Comment:
This `TaskTrackers` looks really confusing to me.

We spawn one in `http_connections` inside we spawn one for `ws_connections` and if it was actual ws connection, we spawn one more for `ws_connections`. 

It would be nice to flatter this in the future. Could you please add TODO?

Reply:
yeah I'm not proud of this one. I was trying to keep the diff as small as possible, but it still got pretty big. I'm gonna brew on this one and see if I can abstract it a bit nicer. I will probably at the very least combine the two task trackers into one because they're redundant.

Reply:
I made the diff bigger :D but it flattens the code quite a bit. WDYT?